Lib/test/test_urllib.py

```diff
@@ -1463,6 +1463,28 @@ def open_spam(self, url):
                 "spam://c:|windows%/:=&?~#+!$,;'@()*[]|/path/"),
                 "//c:|windows%/:=&?~#+!$,;'@()*[]|/path/")
 
+    def test_urlopener_retrieve_file(self):
+        with self.assertWarns(DeprecationWarning):
+            try:
+                fd, tmp_file = tempfile.mkstemp()
```

Comment:
Can't we use `tempfile.NamedTemporaryFile` and context management protocol to avoid handling resources manually?

Reply:
I guess I used it due to Windows permission error in previous CI runs where the created file was not readable but with `mkstemp` it's possible as seen from the docstring. It seems to be a windows only case but it seems to have other problem where the directory could be used by another process as seen in current CI run. Is there a reliable way to generate a temp file for this usecase?

```
    The file is readable and writable only by the creating user ID.
    If the operating system uses permission bits to indicate whether a
    file is executable, the file is executable by no one. The file
    descriptor is not inherited by children of this process.
```

Reply:
I think `support.unlink()` can be used to handle "it's being used by another process" errors under Windows.

Also, the order of `self.addCleanup()` calls is important. You might be bitten by a similar issue like https://github.com/python/cpython/commit/d0f5bab21b9828f4764bd45e4afe24f23c1a55b0.

Reply:
NamedTemporaryFile gives me that the file is not readable. Hence I used temp_dir and mkstemp to generate a more safer version of temp file for this test but I am sure I am not the only one having this problem in the testsuite.

Reply:
I don't think calling `os.close()` just after `tempfile.mkstemp()` makes sense. The temp directory will be cleaned up after the `support.temp_dir()` context manager and `fd` will be effectively invalid.

Reply:
Without calling `os.close(fd)` there will be an open file handle to the tempfile inside the directory and there will be a permission error that file handle is used by another process cannot be cleaned up as `support.temp_dir` context manager exits and due to unclosed fd. I guess Windows doesn't allow deleting directory when there is an open file handle to one of the files inside.

Reply:
Ah, now I understood what you are talking about, and, yes, your analysis is correct. I think the current version of the test is good and reliable enough. Thanks!